### PR TITLE
Recipe App 👨‍💻 👨‍💻 

### PR DESCRIPTION
**Pull request by tsheporamantso**
### On this milestone we implemented the following
- Login page and registration page:
     - Should be built with Devise.
- Recipes list:
     - Should display a list of recipes created by the logged-in user as in the wireframe.
     - Should lead to recipe details.
- Public recipe list:
     - Should display a list of all public recipes ordered by newest as in the wireframe.
     - Should lead to recipe details.
     - If the user is the owner of the recipe, should allow the user to delete it.
- Recipe details:
     - Should display a toggle button that allows for a recipe to be made public or private.
     - If the recipe is public or the user is the owner of the recipe, should display the recipe details as in the wireframe.
     - If the user is the owner of the recipe, should lead to the form that allows the user to add new food.
- Make sure there are no N+1 queries happening.
- Create a navigation menu that allows users to open all of the pages you created.
- Write unit and integration tests

- Food list:
     - Should display a list of food added by the logged-in user as in the wireframe (display also quantity of a given food).
     - Should lead to a form that allows users to add new food.
- General shopping list view:
     - Should show the list of food that is missing for all recipes of the logged-in user (compare the list of food for all recipes with the general food list of that user).
     - Should count the total food items and total price of the missing food.
- Check and correct Rubocop offences
- Check and correct Stylelint errors

### CODE REVIEW

File: app/views/recipes/show.html.erb [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+              <span class="slider"></span>
+            </label>
+          </p>
+        </div>
+      <% end %>
+    </div>
+  </div>
+
+  <div class="btns">
+    <%= button_to "Generate Shopping List", "", class: "add-food-link"%>
```
Comment (Ibizugbe) [inline app/views/recipes/show.html.erb]: Nice job so far👍🏼. I noticed that when the `Generate shopping list` button is clicked, an error is rendered which breaks the app, even though the shopping list is generated. Kindly fix this, as this can lead to a poor user experience. I suggest that when the button is clicked, it should lead the user to the general shopping list view so the user can see the result of his actions😉.

image

File: app/views/recipes/new.html.erb [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    <div>
+      <%= form.label :cooking_time %><br>
+      <%= form.number_field :cooking_time, in: 1..1000, step: 1, autocomplete: "off" %>
+    </div>
+
+    <div>
+      <%= form.label :public %><br>
+      <%= form.check_box :public %>
+    </div>
+
```
Comment (Ibizugbe) [inline app/views/recipes/new.html.erb]: I noticed you added validation to all your forms, which is a good practice😉. The issue currently is that, for all your input fields or forms, whenever a user proceeds to submit with any empty field, it breaks the app. This can lead to a poor user experience. If a field is not provided, I suggest you send an error message to the user on the same page or you can use the `required` HTML attribute that prevents the user from submitting the form automatically😉.

In summary, let's find a way of preventing the user from submitting a form with empty fields, but without breaking the app👍🏼. Please remember to implement this for all your forms😉.

image

**Review by Ibizugbe — CHANGES_REQUESTED** [review]
Hi @tsheporamantso  @Fombi-Favour 👋🏼👋🏼👋🏼, 

request change


Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

## Highlights✅✅✅

- [x] Account management is built using devise👍🏼.
- [x] Unit and integration tests have been written and all tests are passing👍🏼.
- [x] `n+1` concerns have been taken care of👍🏼.
- [x] Nice job implementing the toggle button for making a recipe public👍🏼.
- [x] github flow is used👍🏼.
- [x] Linters are passing👍🏼.


## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by AmaduKamara — APPROVED** [review]
# Status: Approved 🟢 🟢 🟢

Hi Team,

Good job working on the changes highlighted by the previous reviewer 👍🏽

You guys did a great work. Keep it up 👍🏽

Your project is complete!  There is nothing else to say other than... it's time to merge it: ship it:
Congratulations! 🎉 

## Highlights:

✅ You set up Gitflow correctly 👍🏽
✅ The Linter checks passed successfully 👍🏽
✅ Good descriptive PR and README file 👍🏽
✅ Good commit messages 👍🏽
✅ The app works fine and all features were implemented as expected 👍🏽
✅ Implemented tests and all tests passed successfully 👍🏽

Cheers, and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._
```
